core/arch/arm/plat-vexpress/main.c

```diff
@@ -180,3 +188,36 @@ static TEE_Result init_tzc400(void)
 
 service_init(init_tzc400);
 #endif /*CFG_TZC400*/
+
+#if defined(PLATFORM_FLAVOR_qemu_virt)
+int psci_cpu_on(uint32_t core_idx, uint32_t entry, uint32_t context_id __unused)
+{
+	size_t pos = get_core_pos_mpidr(core_idx);
+	uint32_t *sec_entry_addrs = phys_to_virt(SECRAM_BASE, MEM_AREA_IO_SEC);
+	static bool core_is_released[CFG_TEE_CORE_NB_CORE];
+
+	if (!sec_entry_addrs)
+		panic();
+
+	if (!pos || pos >= CFG_TEE_CORE_NB_CORE)
+		return PSCI_RET_INVALID_PARAMETERS;
+
+	DMSG("core pos: %zu: ns_entry %#" PRIx32, pos, entry);
+
+	if (core_is_released[pos]) {
+		EMSG("core %zu already released", pos);
+		return PSCI_RET_DENIED;
+	}
+	core_is_released[pos] = true;
+
+	/* set NS entry addresses of core */
+	ns_entry_addrs[pos] = entry;
+	dsb_ishst();
+
+	sec_entry_addrs[pos] = CFG_TEE_RAM_START;
+	dsb_ishst();
+	sev();
```

Comment:
Would it work with a single call to `dsb_ishst()`? I mean:

```C
        ns_entry_addrs[pos] = entry;
        sec_entry_addrs[pos] = CFG_TEE_RAM_START;
        dsb_ishst();
        sev();
```

Reply:
I guess it would but it's more correct to make sure that the updated value of `sec_entry_addrs[pos]` can't be read before `ns_entry_addrs[pos]`.

The second write will release core from the loop at https://github.com/linaro-swg/bios_qemu_tz_arm/blob/master/bios/entry.S#L139

Reply:
OK I see.